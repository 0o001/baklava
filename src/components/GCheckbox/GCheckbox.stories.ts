import { storiesOf } from '@storybook/vue';
import { withKnobs, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import GCheckbox from './';
import GBox from '../GBox';
import GText from '../GText';

storiesOf('GCheckbox', module)
  .addDecorator(withKnobs)
  .addParameters({ component: GCheckbox })
  .add('Single Select', () => ({
    components: { GCheckbox, GText },
    template: `
      <GCheckbox
        :value="{ id: 1 }"
        v-model="isChecked"
        @change="onChange"
        :disabled="disabled"
        :indeterminate="indeterminate"
      >
        <GText variant="body"> Is Selected: {{ isChecked }} </GText>
      </GCheckbox>
    `,
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
      indeterminate: {
        default: boolean('Indeterminate', false),
      },
    },
    data () {
      return {
        isChecked: false,
      };
    },
    methods: {
      onChange (event) {
        action('Change')(event);
      },
    },
  }))
  .add('Multi Select with object values', () => ({
    components: { GCheckbox, GBox, GText },
    template: `
      <GBox>
        <GBox flex direction="row" :mb="10">
          <GCheckbox :value="value" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="{ id: 2 }" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="{ id: 3 }" v-model="selectedItems" :disabled="disabled"></GCheckbox>
        </GBox>
        <GText variant="body">Selected Items:</GText>
        <ul>
            <li v-for="selected in selectedItems"> {{ selected }} </li>
        </ul>
      </GBox>
    `,
    data () {
      const value = { id: 1 };
      return {
        value,
        selectedItems: [value],
      };
    },
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
    },
  }))
  .add('Multi Select with string values', () => ({
    components: { GCheckbox, GBox, GText },
    template: `
      <GBox>
        <GBox flex direction="row" :mb="10">
          <GCheckbox :value="'value1'" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="'value2'" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="'value3'" v-model="selectedItems" :disabled="disabled"></GCheckbox>
        </GBox>
        <GText variant="body">Selected Items:</GText>
        <ul>
            <li v-for="selected in selectedItems"> {{ selected }} </li>
        </ul>
      </GBox>
    `,
    data () {
      return {
        selectedItems: ['value2'],
      };
    },
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
    },
  }))
  .add('Props', () => ({
    components: { GCheckbox, GText },
    template: `
      <GCheckbox
        :value="{ id: 1 }"
        v-model="isChecked"
        @change="onChange"
        :disabled="disabled"
        :isInvalid="isInvalid"
        color="turquoise-500"
      >
        <GText variant="body"> Is Selected: {{ isChecked }} </GText>
      </GCheckbox>`,
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
      isInvalid: {
        default: boolean('Invalid', false),
      },
    },
    data () {
      return {
        isChecked: false,
      };
    },
    methods: {
      onChange (event) {
        action('Change')(event);
      },
    },
  }))
  .add('Events', () => ({
    components: { GCheckbox },
    template: '<GCheckbox :value="{ id: 1 }" @change="onChange" @click="onClick" />',
    methods: {
      onChange (event) {
        action('Changed')(event);
      },
      onClick (event) {
        action('Clicked')(event);
      },
    },
  }));
