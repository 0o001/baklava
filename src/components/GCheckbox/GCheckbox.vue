<template>
  <GBox
    flex
    direction="row"
    align="center"
  >
    <label class="g-checkbox">
      <input
        type="checkbox"
        class="-input"
        v-bind="$attrs"
        v-on="listeners"
        :value="value"
        :checked="isChecked"
        @change="toggleInput"
      >
      <span
        class="-inner"
        :class="{
          '-disabled': $attrs.disabled,
          '-invalid': isInvalid,
          '-checked': isChecked,
          [`g-bg-${color}`]:isChecked,
          [`g-bg-white`]:!isChecked
        }"
      >
        <GIcon
          size="16px"
          :name="$attrs.indeterminate ? 'minus' : 'check'"
          :stroke="strokeColor"
        />
      </span>
    </label>
    <GBox
      :ml="10"
      v-if="hasSlot"
    >
      <slot />
    </GBox>
  </GBox>
</template>

<script lang="ts">

import GBox from '../GBox/GBox.vue';
import GIcon from '../GIcon/GIcon.vue';

export default {
  name: 'ScCheckbox',
  model: {
    prop: 'modelValue',
    event: 'input',
  },
  components: {
    GBox,
    GIcon,
  },

  props: {
    value: {
      // type: Object | String,
      default: null,
    },
    modelValue: {
      // type: Array | Boolean,
      default: null,
    },
    isInvalid: {
      type: Boolean,
      default: false,
    },
    color: {
      type: String,
      default: 'orange-500',
    },
  },

  data () {
    return {
      isChecked: false,
    };
  },

  computed: {
    hasSlot () {
      return typeof this.$slots.default !== 'undefined';
    },
    listeners () {
      const { change, input, ...restListeners } = this.$listeners;

      return restListeners;
    },
    strokeColor () {
      if (this.$attrs.disabled && this.isChecked) {
        return 'var(--light-grey-500)';
      } else if (this.$attrs.disabled && !this.isChecked) {
        return 'var(--bg-grey-500)';
      } else {
        return 'var(--white)';
      }
    },
    hasSlot () {
      return typeof this.$slots.default !== 'undefined';
    },
  },

  methods: {
    toggleInput (event) {
      this.isChecked = event.target.checked;
      let value;

      if (Array.isArray(this.modelValue)) {
        value = [...this.modelValue];

        if (this.isChecked) {
          value.push(this.value);
        } else {
          value = value
            .filter(val => val.id
              ? val.id !== this.value.id
              : val !== this.value);
        }
      } else {
        value = this.isChecked;
      }

      this.$emit('change', { event, value: this.value, checked: this.isChecked });
      this.$emit('input', value);
    },
  },

  watch: {
    modelValue: {
      handler (newModelValue) {
        if (Array.isArray(newModelValue)) {
          if (newModelValue.indexOf(this.value) !== -1) {
            this.isChecked = true;
          } else if (newModelValue.length === 0) {
            this.isChecked = false;
          }
        } else {
          this.isChecked = newModelValue;
        }
      },
      immediate: true,
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-checkbox {
    cursor: pointer;
    display: inline-block;
    position: relative;
    width: var(--spacing-16);
    height: var(--spacing-16);

    .-input {
      cursor: pointer;
      position: absolute;
      opacity: 0;
      top: var(--spacing-0);
      left: var(--spacing-0);
      width: var(--spacing-16);
      height: var(--spacing-16);
    }

    .-inner {
      text-align: center;
      width: var(--size-full);
      height: var(--size-full);
      border-radius: var(--radius-sm);
      display: flex;

      &.-checked {
        border: solid 1px transparent !important;
      }

      &.-disabled {
        background-color: var(--bg-grey-500);
        border: solid 1px var(--mid-grey-800) !important;
      }

      &.-invalid {
        border: solid 1px var(--red-500) !important;
      }

      &:not(.-invalid) {
        border: solid 1px var(--mid-grey-800);
      }
    }
  }
</style>
