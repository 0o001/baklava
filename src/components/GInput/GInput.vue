<template>
  <div :class="wrapperClass">
    <slot name="left-addon" />
    <GFieldWrapper
      class="body"
      :is-value="!!value"
      :icon-position="iconPosition"
      :is-outline-label="isOutlineLabel"
      :feedback="feedback"
      :label="label"
      :icon="icon"
      :success="success"
      :error="error"
      :disable="disable"
      :size="size"
      @icon-clicked="$emit('icon-clicked', $event)"
    >
      <!--
      @slot You can put select, button and box in this field
      -->
      <input
        :type="type"
        @input="onInput"
        v-on="getListeners"
        v-bind="getAttrs"
        ref="input"
        :disabled="disable"
        :class="labelClass"
        :value="getValue"
      >
      <!--
      @slot You can put select, button and box in this field
      -->
    </GFieldWrapper>
    <slot name="right-addon" />
  </div>
</template>

<script lang="ts">

import ClickOutside from '../../directives/ClickOutside';
import GFieldWrapper from '../GFieldWrapper/GFieldWrapper.vue';

export default {
  name: 'GInput',
  components: {
    GFieldWrapper,
  },
  directives: {
    ClickOutside,
  },
  props: {
    value: {
      type: [String, Number, Date],
      default: () => null,
    },
    type: {
      default: 'text',
      type: String,
    },
    isOutlineLabel: {
      default: false,
      type: Boolean,
    },
    success: {
      default: false,
      type: Boolean,
    },
    error: {
      default: false,
      type: Boolean,
    },
    disable: {
      default: false,
      type: Boolean,
    },
    label: {
      default: '',
      type: String,
    },
    feedback: {
      default: '',
      type: String,
    },
    icon: {
      default: '',
      type: String,
    },
    iconPosition: {
      default: '',
      type: String,
    },
    size: {
      default: 'big',
      type: String,
    },
  },
  computed: {
    getAttrs () {
      const { value, ...rest } = this.$attrs;
      return rest;
    },
    getListeners () {
      const { input, ...rest } = this.$listeners;
      return rest;
    },
    wrapperClass () {
      return {
        'g-input': true,
        'left-addon': !!this.$slots['left-addon'],
        'right-addon': !!this.$slots['right-addon'],
      };
    },
    labelClass () {
      let classValue;
      if (this.error) return 'g-text-red-500';
      if (this.success) return 'g-text-green-500';
      if (this.disable) return 'g-text-dark-grey-500';
      return classValue;
    },
    getValue () {
      return this.value;
    },
  },
  methods: {
    onInput (e: any) {
      this.$emit('input', e.target.value);
    },
  },
};
</script>

<style lang="scss" scoped>
.g-input {
  &.right-addon {
    display: flex;
    > .body {
      width: 100%;
      ::v-deep .g-field-wrapper {
        .content {
          border-top-right-radius: 0;
          border-bottom-right-radius: 0;
        }
      }
    }
    ::v-deep > .g-input {
      .g-field-wrapper .content {
        border-left: 0;
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        border-top-right-radius: var(--radius-md);
        border-bottom-right-radius: var(--radius-md);
      }
    }
    ::v-deep > .g-button {
      z-index: 2;
      margin-left: -1px;
      height: fit-content;
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
    }
    ::v-deep > .g-native-select {
      min-width: 20%;
      .g-field-wrapper .content {
        border-left: 0;
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        border-top-right-radius: var(--radius-md);
        border-bottom-right-radius: var(--radius-md);
      }
    }
  }
  &.left-addon {
    display: flex;
    > .body {
      width: 100%;
      ::v-deep .g-field-wrapper {
        .content {
          border-top-left-radius: 0;
          border-bottom-left-radius: 0;
        }
      }
    }
    ::v-deep > .g-input {
      .g-field-wrapper .content {
        border-right: 0;
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
        border-top-left-radius: var(--radius-md);
        border-bottom-left-radius: var(--radius-md);
      }
    }
    ::v-deep > .g-button {
      height: fit-content;
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }
    ::v-deep > .g-native-select {
      min-width: 20%;
      .g-field-wrapper .content {
        border-right: 0;
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
        border-top-left-radius: var(--radius-md);
        border-bottom-left-radius: var(--radius-md);
      }
    }
  }
}
</style>
