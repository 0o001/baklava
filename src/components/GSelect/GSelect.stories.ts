import { storiesOf } from '@storybook/vue';
import GSelect from './GSelect.vue';
import { action } from '@storybook/addon-actions';

storiesOf('GSelect', module)
  .addParameters({ component: GSelect })
  .add('Default', () => ({
    components: { GSelect },
    data () {
      return {
        value1: ['Value2'],
        value2: '',
        value3: 'Value2',
        value4: 'Value2',
        value5: '',
        value6: 'Value2',
        value7: '',
        value8: [],
      };
    },
    props: {
      options: {
        default: [
          {
            value: 'Value1',
            text: 'Text Option 1 with i',
          },
          {
            value: 'Value2',
            text: 'Text Option 2 with i',
          },
          {
            value: 'Value3',
            text: 'Text Option 3 with i',
            disabled: true,
          },
        ],
      },
    },
    methods: {
      onSearchChange (text: string) {
        action('On Search Change')(text);
      },
      onChange (value: string | string[]) {
        action('On Change')(value);
      },
      onBlur (value: string | string[]) {
        action('On blur')(value);
      },
    },
    template: `
    <div style="margin: 40px 10vh 10vh">
      <GSelect class="g-mb-20" :options="options" v-model="value1" label="Email" @onChange="onChange" placeholder="Placeholder" isCheckbox @blur="onBlur"/>
      <GSelect class="g-mb-20" :options="options" v-model="value2" label="Email" @onChange="onChange" placeholder="Placeholder" isSearch @onSearchChange="onSearchChange" feedback="feedback" isOutlineLabel/>
      <GSelect class="g-mb-20" :options="options" v-model="value3" label="Email" />
      <GSelect class="g-mb-20" :options="options" v-model="value4" label="Small" size="small" />
      <GSelect class="g-mb-20" :options="options" v-model="value5" label="Middle" size="middle" placeholder="Placeholder" />
      <GSelect class="g-mb-20" :options="options" v-model="value6" label="Big" size="big" />
      <GSelect class="g-mb-20" :options="options" v-model="value7" placeholder="Email" isBorderless />
      <div style="width: 150px">
        <GSelect class="g-mb-20" :options="options" v-model="value8" label="Email" @onChange="onChange" placeholder="Placeholder" isCheckbox isOutlineLabel/>
      </div>
    </div>`,
  }));
