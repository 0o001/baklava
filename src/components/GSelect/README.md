# GSelect

## Props

| Name                 | Type                    | Description | Default   |
| -------------------- | ----------------------- | ----------- | --------- |
| `is-outline-label`   | `Boolean`               |             | `false`   |
| `is-borderless`      | `Boolean`               |             | `false`   |
| `is-checkbox`        | `Boolean`               |             | `false`   |
| `is-search`          | `Boolean`               |             | `false`   |
| `success`            | `Boolean`               |             | `false`   |
| `error`              | `Boolean`               |             | `false`   |
| `disable`            | `Boolean`               |             | `false`   |
| `label`              | `String`                |             | `""`      |
| `placeholder`        | `String`                |             | `""`      |
| `search-placeholder` | `String`                |             | `"Arama"` |
| `feedback`           | `String`                |             | `""`      |
| `v-model`            | `String` &#124; `Array` |             | `""`      |
| `options` *required* | `Array`                 |             |           |
| `size`               | `String`                |             | `"big"`   |

## Data

| Name               | Type      | Description | Initial value |
| ------------------ | --------- | ----------- | ------------- |
| `isOptionsVisible` | `boolean` |             | `false`       |
| `searchText`       | `string`  |             | `""`          |

## Computed Properties

| Name              | Type      | Description                                                                           |
| ----------------- | --------- | ------------------------------------------------------------------------------------- |
| `filteredOptions` | `unknown` | **Dependencies:** `options`, `searchText`                                             |
| `icon`            | `unknown` | **Dependencies:** `isOptionsVisible`                                                  |
| `getValue`        | `object`  | **Dependencies:** `value`                                                             |
| `getLabel`        | `unknown` | **Dependencies:** `options`, `value`, `isOutlineLabel`, `isBorderless`, `placeholder` |
| `wrapperClass`    | `object`  | **Dependencies:** `size`, `isBorderless`                                              |
| `contentClass`    | `object`  | **Dependencies:** `isOptionsVisible`                                                  |

## Events

| Name             | Description                                              |
| ---------------- | -------------------------------------------------------- |
| `onSearchChange` | <br/>**Arguments**<br/><ul><li>**`text: any`**</li></ul> |
| `input`          | <br/>**Arguments**<br/><ul><li>**`item: any`**</li></ul> |
| `onChange`       | <br/>**Arguments**<br/><ul><li>**`item: any`**</li></ul> |

## Methods

### isSelected()

**Syntax**

```typescript
isSelected(item: unknown): Boolean
```

### onSearchChange()

**Syntax**

```typescript
onSearchChange(text: string | number): void
```

### clickItem()

**Syntax**

```typescript
clickItem(item: string | number): void
```

### clickCheckbox()

**Syntax**

```typescript
clickCheckbox(item: string | number): void
```

