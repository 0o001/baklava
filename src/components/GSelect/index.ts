import GSelect from './GSelect.vue';

export default GSelect;
