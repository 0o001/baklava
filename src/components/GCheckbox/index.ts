import GCheckbox from './GCheckbox.vue';

export default GCheckbox;
