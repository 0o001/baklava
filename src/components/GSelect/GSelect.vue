<template>
  <div :class="wrapperClass">
    <GFieldWrapper
      :is-active-content.sync="isOptionsVisible"
      :is-value="Array.isArray(value)? !!value.length: !!value"
      :is-outline-label="isOutlineLabel"
      :feedback="feedback"
      :label="label"
      :icon="icon"
      :success="success"
      :error="error"
      :disable="disable"
    >
      <p
        class="valueText"
        v-text="getLabel"
      />
      <div :class="contentClass">
        <GBox
          :mb="15"
          v-if="isSearch"
        >
          <GInput
            :placeholder="searchPlaceholder"
            icon="search"
            @click.stop
            @input="onSearchChange"
            size="small"
            v-model="searchText"
          />
        </GBox>
        <GBox
          flex
          align="start"
          class="item"
          v-for="item in filteredOptions"
          :key="item.value"
          @click.stop="clickItem(item.value)"
        >
          <GCheckbox
            v-if="isCheckbox"
            :value="{ id: 1 }"
            :model-value="isSelected(item)"
            class="g-mr-4"
          />
          <span
            class="text"
            v-text="item.text"
          />
        </GBox>
      </div>
    </GFieldWrapper>
  </div>
</template>

<script lang="ts">

import GFieldWrapper from '../GFieldWrapper/GFieldWrapper.vue';
import GInput from '../GInput/GInput.vue';
import GCheckbox from '../GCheckbox/GCheckbox.vue';
import GBox from '../GBox/GBox.vue';

export default {
  name: 'GSelect',
  components: {
    GFieldWrapper,
    GInput,
    GCheckbox,
    GBox,
  },
  props: {
    isOutlineLabel: {
      default: false,
      type: Boolean,
    },
    isBorderless: {
      default: false,
      type: Boolean,
    },
    isCheckbox: {
      default: false,
      type: Boolean,
    },
    isSearch: {
      default: false,
      type: Boolean,
    },
    success: {
      default: false,
      type: Boolean,
    },
    error: {
      default: false,
      type: Boolean,
    },
    disable: {
      default: false,
      type: Boolean,
    },
    label: {
      default: '',
      type: String,
    },
    placeholder: {
      default: '',
      type: String,
    },
    searchPlaceholder: {
      default: 'Arama',
      type: String,
    },
    feedback: {
      default: '',
      type: String,
    },
    value: {
      default: '',
      type: [String, Array],
    },
    options: {
      required: true,
      type: Array,
    },
    size: {
      /* small, middle, big */
      default: 'big',
      type: String,
    },
  },
  data () {
    return {
      isOptionsVisible: false,
      searchText: '',
    };
  },
  methods: {
    isSelected (item): Boolean {
      if (Array.isArray(this.value)) {
        return this.value.includes(item.value);
      }
      return this.value === item.value;
    },
    onSearchChange (text: string | number) {
      this.$emit('onSearchChange', text);
    },
    clickItem (item: string | number) {
      if (this.isCheckbox) {
        return this.clickCheckbox(item);
      }
      this.isOptionsVisible = false;
      this.$emit('input', item);
      this.$emit('onChange', item);
    },
    clickCheckbox (item: string | number) {
      const array = this.value.slice();
      const i = array.findIndex(x => x === item);
      if (i >= 0) {
        array.splice(i, 1);
      } else {
        array.push(String(item));
      }
      this.$emit('onChange', array);
      this.$emit('input', array);
    },
  },
  computed: {
    filteredOptions () {
      return this.options.filter(opt =>
        (opt.text || '').toLocaleLowerCase('TR')
          .includes((this.searchText || '').toLocaleLowerCase('TR')));
    },
    icon  () {
      return this.isOptionsVisible ? 'chevron-up' : 'chevron-down';
    },
    getValue () {
      return this.value;
    },
    getLabel () {
      const selectedOptions = this.options.filter(opt => {
        if (Array.isArray(this.value)) {
          return this.value.includes(opt.value);
        }
        return opt.value === this.value;
      });
      if (!selectedOptions.length) {
        return (this.isOutlineLabel || this.isBorderless) ? this.placeholder : '';
      }

      return selectedOptions.map(opt => opt.text).join(', ');
    },
    wrapperClass () {
      return {
        'g-select': true,
        [this.size]: true,
        '-borderless': this.isBorderless,
      };
    },
    contentClass () {
      return {
        's-content': true,
        'g-d-none': !this.isOptionsVisible,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
.g-select {
  ::v-deep .g-field-wrapper {
    &.-active {
     .content {
        border-bottom: none;
        border-radius: 0;
        border-top-left-radius: var(--radius-md);
        border-top-right-radius: var(--radius-md);
        border-color: var(--orange-500);
      }
    }
    .content {
      select {
        -webkit-appearance: none;
        z-index: 2;
        color: var(--main-grey-500);
      }
    }
    &.big .content{
      height: 38px;
    }
    &.small .content{
      height: 32px;
    }
    .g-field-wrapper .content {
      background:var(--bg-grey-500);
      border: 1px solid var(--mid-grey-500);
      border-radius: var(--radius-md);
    }
  }
  .item{
    padding-bottom: 10px;
    border-bottom: 1px solid #d5d9e1;
    margin-bottom: 10px;
    line-height: 1.4;
  }
  .item:last-child {
    border: none;
    padding: 0;
    margin: 0;
  }
  &.-borderless {
    ::v-deep .g-field-wrapper {
      background: var(--bg-grey-500);
      border-radius: var(--radius-md);
      .content {
        border: none;
        color: var(--main-grey-500);
        .s-content {
          border: 1px solid var(--orange-500);
          border-radius: var(--radius-md);
          margin-top: 15px;
          .text{
            margin-left: 12px;
          }
        }
      }
    }
  }
  .s-content {
    width: -webkit-fill-available;
    background: white;
    padding: 15px;
    border: 1px solid var(--mid-grey-800);
    border-radius: var(--radius-md);
    position: absolute;
    top: 34px;
    z-index: 3;
    margin: -1px;
    &:not(.d-none) {
      border-top: none;
      border-radius: 0;
      border-bottom-left-radius: var(--radius-md);
      border-bottom-right-radius: var(--radius-md);
      border-color: var(--orange-500);
    }
  }
  .text {
    font-family: var(--font-family-base);
    font-size: var(--font-size-14);
    color: var(--dark-grey-500);
    width: 100%;
    &:hover {
      color: var(--main-grey-500);
      cursor: pointer;
    }
  }
  .valueText {
    font-family: var(--font-family-base);
    font-size: var(--font-size-14);
    color: var(--main-grey-500);
    padding: 13px 30px 13px 13px;
    display: block;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
  &.small {
    ::v-deep .g-field-wrapper {
      label {
        font-size: 8px;
      }
      .content {
        height: 24px;
        .valueText {
          font-size: 12px;
          padding: 6px;
        }
      }
      .g-icon {
        margin-top: 0px;
      }
    }
    .s-content {
      top: 25px;
      padding: 8px;
    }
    .item{
      margin-bottom: 5px;
      padding-bottom: 5px;
      &:last-child {
        margin: 0;
        padding: 0;
      }
    }
  }
  &.middle {
    ::v-deep .g-field-wrapper {
      label {
        font-size: 12px;
      }
      .content {
        height: 32px;
        .valueText {
          font-size: 14px;
          padding: 10px;
        }
      }
    }
    .s-content {
      top: 30px;
      padding: 10px;
    }
    .item{
      margin-bottom: 10px;
      padding-bottom: 10px;
      &:last-child {
        margin: 0;
        padding: 0;
      }
    }
  }
  .s-content{
    max-height: 200px;
    overflow: auto;
  }
}
</style>
