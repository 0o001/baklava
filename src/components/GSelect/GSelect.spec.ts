import { mount, Wrapper, shallowMount } from '@vue/test-utils';
import GSelect from '.';
import GCheckbox from '../GCheckbox';
import GInput from '../GInput';

describe('GSelect', () => {
  let wrapper: Wrapper<any>;

  it('should match snapshot when show is prop is given as false ', () => {
    wrapper = mount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render given default when options prop set', () => {
    wrapper = shallowMount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
      },
    });
    expect(wrapper.vm.icon).toEqual('chevron-down');
    expect(wrapper.vm.getValue).toEqual('');
    expect(wrapper.vm.wrapperClass).toEqual({
      '-borderless': false,
      big: true,
      'g-select': true,
    });
    expect(wrapper.vm.contentClass).toEqual({
      'g-d-none': true,
      's-content': true,
    });
  });

  it('should render given default when defaultValue prop set', () => {
    wrapper = shallowMount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        value: 'Value2',
      },
    });
    expect(wrapper.vm.getValue).toEqual('Value2');
  });

  it('should render given default when size prop set', () => {
    wrapper = shallowMount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        size: 'small',
      },
    });
    expect(wrapper.vm.wrapperClass).toEqual({
      '-borderless': false,
      small: true,
      'g-select': true,
    });
  });

  it('should render given default when isBorderless and placeholder prop set', () => {
    wrapper = shallowMount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        placeholder: 'placeholder',
        isBorderless: true,
      },
    });
    expect(wrapper.vm.getLabel).toEqual('placeholder');
    expect(wrapper.vm.wrapperClass).toEqual({
      '-borderless': true,
      big: true,
      'g-select': true,
    });
  });

  it('the correct class and icon appears when the GSelect is clicked', async () => {
    wrapper = mount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        placeholder: 'placeholder',
        isBorderless: true,
      },
    });
    wrapper.findAll('.content').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.icon).toEqual('chevron-up');
    expect(wrapper.vm.contentClass).toEqual({
      'g-d-none': false,
      's-content': true,
    });
  });

  it('should render checkbox when isCheckbox prop set', () => {
    wrapper = shallowMount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        isCheckbox: true,
      },
    });

    expect(wrapper.findAllComponents(GCheckbox).length).toEqual(2);
    expect(wrapper.findAll('.text').length).toEqual(2);
    expect(
      wrapper
        .findAll('.text')
        .at(1)
        .text(),
    ).toEqual('Text2');
  });

  it('should render checkbox when isSearch prop set', () => {
    wrapper = shallowMount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        isSearch: true,
      },
    });

    expect(wrapper.findAllComponents(GInput).length).toEqual(1);
  });

  it('should emit blur event if select is closed', async () => {
    // given
    wrapper = mount(GSelect, {
      propsData: {
        options: [
          { value: 'Value1', text: 'Text1' },
          { value: 'Value2', text: 'Text2' },
        ],
        placeholder: 'placeholder',
        isBorderless: true,
      },
    });

    wrapper.find('.content').trigger('click'); // open select
    await wrapper.vm.$nextTick();

    // when
    wrapper.find('.content').trigger('click'); // close select
    await wrapper.vm.$nextTick();

    // then
    expect(wrapper.emitted().blur).toBeTruthy();
  });
});
