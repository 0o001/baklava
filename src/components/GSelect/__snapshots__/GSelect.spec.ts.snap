// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GSelect should match snapshot when show is prop is given as false  1`] = `
<div
  class="g-select big"
>
  <div>
    <div
      class="g-field-wrapper -borderline big"
    >
      <!---->
       
      <div
        class="content"
      >
        <p
          class="valueText"
        />
         
        <div
          class="s-content g-d-none"
        >
          <!---->
           
          <div
            class="item g-d-flex g-d-align-start"
          >
            <!---->
             
            <span
              class="text"
            >
              Text1
            </span>
          </div>
          <div
            class="item g-d-flex g-d-align-start"
          >
            <!---->
             
            <span
              class="text"
            >
              Text2
            </span>
          </div>
        </div>
         
        <svg
          class="g-icon feather feather--chevron-down"
          dataName="chevron-down"
          dataTags=""
          dataType="chevron-down"
          fill="none"
          height="24px"
          stroke="var(--light-grey-500)"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
       
      <!---->
    </div>
  </div>
</div>
`;
